src/server/services/email-verification.service.ts

```diff
@@ -0,0 +1,139 @@
+import { randomBytes } from 'crypto';
+import { dbRead, dbWrite } from '~/server/db/client';
+import { env } from '~/env/server';
+import { emailVerificationEmail } from '~/server/email/templates/emailVerification.email';
+import { throwBadRequestError, throwNotFoundError } from '~/server/utils/errorHandling';
+import { REDIS_KEYS, redis } from '~/server/redis/client';
+import { invalidateSession } from '~/server/utils/session-helpers';
+
+const EMAIL_VERIFICATION_EXPIRY = 15 * 60; // 15 minutes in seconds
+
+export async function generateEmailVerificationToken(userId: number, newEmail: string) {
+  const token = randomBytes(32).toString('hex');
+
+  // Store the verification data in Redis with expiry
+  await redis.set(
+    `${REDIS_KEYS.EMAIL_VERIFICATION}:${token}`,
+    JSON.stringify({
+      userId,
+      newEmail,
+      createdAt: new Date().toISOString(),
+    }),
+    {
+      EX: EMAIL_VERIFICATION_EXPIRY, // 15 minutes
+    }
+  );
+
+  return token;
+}
+
+export async function validateEmailChangeToken(token: string) {
+  const data = await redis.get(`${REDIS_KEYS.EMAIL_VERIFICATION}:${token}`);
+
+  if (!data) {
+    throw throwBadRequestError('Invalid or expired verification token');
+  }
+
+  const verificationData = JSON.parse(data);
```

Comment:
Probably good idea to wrap this in try/catch to handle parsing errors gracefully